node/impl/full/multisig.go

```diff
@@ -0,0 +1,271 @@
+	}
+
+	// wait for it to get mined into a block
+	wait, err := a.StateAPI.StateWaitMsg(ctx, smsg.Cid())
+	if err != nil {
+		return api.UndefAddressPair, err
+	}
+
+	// check it executed successfully
+	if wait.Receipt.ExitCode != 0 {
+		fmt.Println("actor creation failed!")
+		return api.UndefAddressPair, err
+	}
+
+	// get address of newly created miner
+
+	var execreturn init_.ExecReturn
+	if err := execreturn.UnmarshalCBOR(bytes.NewReader(wait.Receipt.Return)); err != nil {
+		return api.UndefAddressPair, err
+	}
+	fmt.Println("Created new multisig: ", execreturn.IDAddress, execreturn.RobustAddress)
+
+	return api.AddressPair{
+		IDAddress:     execreturn.IDAddress,
+		RobustAddress: execreturn.RobustAddress,
+	}, nil
+}
+
+func (a *MsigAPI) MsigPropose(ctx context.Context, msig address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params string) (api.MsigProposeOutput, error) {
+
+	if msig == address.Undef {
+		return api.UndefMsigProposalOutput, xerrors.Errorf("must provide a multisig address for proposal")
+	}
+
+	if to == address.Undef {
+		return api.UndefMsigProposalOutput, xerrors.Errorf("must provide a target address for proposal")
+	}
+
+	if !types.BigPos(amt) {
+		return api.UndefMsigProposalOutput, xerrors.Errorf("must provide a positive amount for proposed send")
+	}
+
+	if src == address.Undef {
+		defaddr, err := a.WalletAPI.WalletDefaultAddress(ctx)
+		if err != nil {
+			return api.UndefMsigProposalOutput, err
+		}
+		src = defaddr
+	}
+
+	paramBytes, err := hex.DecodeString(params)
+	if err != nil {
+		return api.UndefMsigProposalOutput, err
+	}
+
+	enc, actErr := actors.SerializeParams(&samsig.ProposeParams{
+		To:     to,
+		Value:  amt,
+		Method: abi.MethodNum(method),
+		Params: paramBytes,
+	})
+	if actErr != nil {
+		return api.UndefMsigProposalOutput, actErr
+	}
+
+	msg := &types.Message{
+		To:       msig,
+		From:     src,
+		Value:    types.NewInt(0),
+		Method:   builtin.MethodsMultisig.Propose,
+		Params:   enc,
+		GasLimit: 100000,
+		GasPrice: types.NewInt(1),
+	}
+
+	smsg, err := a.MpoolAPI.MpoolPushMessage(ctx, msg)
+	if err != nil {
+		return api.UndefMsigProposalOutput, nil
+	}
+
+	fmt.Println("send proposal in message: ", smsg.Cid())
+
+	wait, err := a.StateAPI.StateWaitMsg(ctx, smsg.Cid())
+	if err != nil {
+		return api.UndefMsigProposalOutput, nil
+	}
+
+	if wait.Receipt.ExitCode != 0 {
+		return api.UndefMsigProposalOutput, fmt.Errorf("proposal returned exit %d", wait.Receipt.ExitCode)
+	}
+
+	_, v, err := cbg.CborReadHeader(bytes.NewReader(wait.Receipt.Return))
+	if err != nil {
+		return api.UndefMsigProposalOutput, nil
+	}
+
+	return api.MsigProposeOutput{
+		MessageCid:    smsg.Cid(),
+		TransactionId: v,
+	}, nil
+}
+
+func (a *MsigAPI) MsigApprove(ctx context.Context, msig address.Address, txId uint64, proposer address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params string) (cid.Cid, error) {
```

Comment:
Adding a `Cancel` method would be great for the sake of completeness. AFAIK, it's just `Approve` with a different method code.

Reply:
I'm going to need to look into how to Cancel a proposal. Maybe you can help me with that if you know how?